### DISCUSSION TITLE
Routing issues with github pages

### DISCUSSION THREAD

**Discussion by cheesesamwich** [community/community#131827] [Programming Help] [answered]
### Select Topic Area

Question

### Body

Hey, sorry if this has been asked before but i couldn't find a solid answer. I'm trying to implement my own static routing system for my personal site- it works if i locally host it but as soon as i host it on github pages it returns a 404 with any routes in the url. 
This is my code
```
import React from 'react';
import ReactDOM from 'react-dom/client';
import { App } from "./App.tsx";

function Route()
{
  const currentLocation = new URL(window.location.href).pathname.substring(1);
  if(subRoutes[currentLocation])
  {
    return subRoutes[currentLocation]();
  }
  return <App/>;
}

const subRoutes = {
    "test": () => <img src={"https://samwich.dev/assets/rupert.jpg"}/>,
};

const root = ReactDOM.createRoot(document.getElementById('root'));
root.render(
  <React.StrictMode>
    <Route/>
  </React.StrictMode>
);

```
![image](https://github.com/community/community/assets/149597648/ffcd3b11-861d-4cf6-924f-de103ca332ab)

I saw some mentions of a 404.html, do i need to implement my router there too for it to function correctly?

**Comment by paiineli** [comment]
Hi,

It sounds like you're running into an issue with routing on GitHub Pages. When hosting on GitHub Pages, any direct navigation to a route other than the root will result in a 404 error because GitHub Pages only serves the index.html file for the root path.

To fix this, you can implement a couple of strategies:

Use Hash Routing: This is the simplest solution. Instead of using regular routing, you can use hash-based routing (using # in your URLs). This way, the URL will not trigger a page refresh, and GitHub Pages will not return a 404.

Example:

const currentLocation = window.location.hash.substring(1);

Implement a 404.html Page: If you want to use regular routing, you can create a 404.html page that redirects to your main page. However, you won't be able to handle routes directly in this file.

Use a Static Site Generator: If your site becomes more complex, consider using a static site generator like Gatsby or Next.js, which can help manage routing and build for deployment on GitHub Pages.

In summary, using hash routing is likely the easiest fix for your current setup. If you have further questions, feel free to ask!

Best,

**Comment by Sicilianii** [comment]
Had the same problem, here's the link
https://stackoverflow.com/questions/11577147/how-to-fix-http-404-on-github-pages

If you find this as helpful, please mark this comment as an answer :)

**Comment by LiteBrite82** [comment]
Hi @cheesesamwich 

Great question!!! Im going to send this over to `Programming Help` to get more exposure to the community.

**Comment by GreeningSiren** [comment, marked as answer]
Hello,
You can check [this](https://github.com/rafgraph/spa-github-pages) guide to use react-router with GitHub pages.
It implements a 404.html page, that "tricks" GitHub pages into loading the page.

If you have any further questions, feel free to message me at any time!
Hope you have a nice day,
GreeningSiren

**Reply by cheesesamwich** [reply]
Thank you so much! Eventually i got it to work by swapping to react router and implementing the 404.html and index.html script in the link you provided

**Reply by GreeningSiren** [reply]
If you got that working, please set my answer as the answer to your problem with the button
"Mark as answer"

**Comment by cheesesamwich** [comment]


**Reply by GreeningSiren** [reply]
I think [this](https://stackoverflow.com/questions/53387913/how-to-separate-css-rules-for-files-in-react) post could help you resolve your problem

**Reply by GreeningSiren** [reply]
React bundles the css into one file, so using specific css on the `body` tag will change to all the pages
